File: Bench/single_run_cuda.py
import onnxruntime as ort
import numpy as np
import os

# ================= CONFIG =================
MODEL_PATH = "Models/yolo12n_op12.onnx"
IMG_SHAPE = (3, 128, 128)
BATCH_SIZE = 1

# ================= INPUT GENERATION =================
def generate_input(batch_size):
    np.random.seed(42)
    return np.random.rand(batch_size, *IMG_SHAPE).astype(np.float32)

# ================= SESSION CREATION (GPU) =================
def create_optimized_cuda_session(model_path):
    so = ort.SessionOptions()
    
    # ---------------- OPTIMIZATIONS ----------------
    # Enable extended graph optimizations (fusions, constant folding, etc.)
    so.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_EXTENDED

    # Enable CPU memory arena (even with GPU EP, useful for CPU fallback)
    so.enable_cpu_mem_arena = True

    # Enable memory pattern optimization (reuses memory buffers)
    so.enable_mem_pattern = True

    # Intra-op threads (affects CPU kernels, useful if fallback occurs)
    so.intra_op_num_threads = 2

    # Inter-op threads (affects parallel execution of independent nodes)
    so.inter_op_num_threads = 2

    # ---------------- END OPTIMIZATIONS ----------------
    
    # Use CUDAExecutionProvider first, fallback to CPUExecutionProvider
    providers = ["CUDAExecutionProvider", "CPUExecutionProvider"]

    return ort.InferenceSession(model_path, sess_options=so, providers=providers)

# ================= INFERENCE =================
def run_single_inference(session, input_data):
    input_name = session.get_inputs()[0].name
    outputs = session.run(None, {input_name: input_data})
    print("Single GPU inference completed")
    return outputs

# ================= MAIN =================
def main():
    if not os.path.exists(MODEL_PATH):
        print(f"Error: Model not found at {MODEL_PATH}")
        return

    print(f"Loading model: {MODEL_PATH} ({os.path.getsize(MODEL_PATH)/1024/1024:.1f} MB)")
    session = create_optimized_cuda_session(MODEL_PATH)

    input_data = generate_input(BATCH_SIZE)
    print(f"Running single GPU inference with input shape: {input_data.shape}")
    run_single_inference(session, input_data)

if __name__ == "__main__":
    main()

File: Bench/single_run_cuda.cpp
#include <iostream>
#include <vector>
#include <random>
#include <onnxruntime/core/session/onnxruntime_cxx_api.h>

// Check for provider availability (macros usually defined by build system)
// For 1.6.0 we use the C-API struct or specific helper
#include <onnxruntime/core/session/onnxruntime_c_api.h>

const std::string MODEL_PATH = "Models/yolo12n_op12.onnx";
const int BATCH_SIZE = 1;
const int IMG_SIZE = 128;

int main() {
    Ort::Env env(ORT_LOGGING_LEVEL_WARNING, "SingleRunCUDA");
    Ort::SessionOptions session_options;

    // Optimizations
    session_options.SetGraphOptimizationLevel(GraphOptimizationLevel::ORT_ENABLE_EXTENDED);
    session_options.SetGraphOptimizationLevel(GraphOptimizationLevel::ORT_ENABLE_ALL);
    session_options.SetOptimizedModelFilePath("optimized_cuda.onnx");
    session_options.EnableCpuMemArena();
    session_options.EnableMemPattern();
    session_options.SetIntraOpNumThreads(4);
    session_options.SetInterOpNumThreads(4);

    // Append CUDA Provider
    // Note: ORT 1.6.0 C++ API uses OrtCUDAProviderOptions
    try {
        OrtCUDAProviderOptions cuda_options{};
        cuda_options.device_id = 0;

        session_options.AppendExecutionProvider_CUDA(cuda_options);
        std::cout << "CUDA Provider Appended." << std::endl;
    } catch (...) {
        std::cerr << "Warning: Could not append CUDA provider. Is ORT built with CUDA?" << std::endl;
    }

    try {
        std::cout << "Loading model: " << MODEL_PATH << std::endl;
        Ort::Session session(env, MODEL_PATH.c_str(), session_options);

        // Data Prep
        std::vector<int64_t> input_dims = {BATCH_SIZE, 3, IMG_SIZE, IMG_SIZE};
        size_t input_size = BATCH_SIZE * 3 * IMG_SIZE * IMG_SIZE;
        std::vector<float> input_data(input_size);
        
        std::mt19937 gen(42);
        std::uniform_real_distribution<float> dis(0.0f, 1.0f);
        for(auto& v : input_data) v = dis(gen);

        Ort::AllocatorWithDefaultOptions allocator;
        std::string input_name_str = session.GetInputName(0, allocator);
        const char* input_names[] = { input_name_str.c_str() };
        
        std::string output_name_str = session.GetOutputName(0, allocator);
        const char* output_names[] = { output_name_str.c_str() };

        auto memory_info = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);
        Ort::Value input_tensor = Ort::Value::CreateTensor<float>(
            memory_info, input_data.data(), input_size, input_dims.data(), input_dims.size()
        );

        std::cout << "Running single GPU inference..." << std::endl;
        session.Run(Ort::RunOptions{nullptr}, input_names, &input_tensor, 1, output_names, 1);
        std::cout << "Single GPU inference completed." << std::endl;

    } catch (const Ort::Exception& e) {
        std::cerr << "ORT Error: " << e.what() << std::endl;
        return -1;
    }

    return 0;
}